// @wixc3/react-board/dist/react-error-handled-render.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.reactErrorHandledRendering = void 0;
const react_1 = __importDefault(require("react"));
const react_dom_1 = __importDefault(require("react-dom"));
const client_1 = __importDefault(require("react-dom/client"));
const reactRootByContainer = new WeakMap();
const reactErrorHandledRendering = async (element, container) => {
    if (client_1.default.createRoot) {
        // react 18+
        const reactRoot = reactRootByContainer.get(container) || client_1.default.createRoot(container);
        reactRootByContainer.set(container, reactRoot);
        await new Promise((resolve, reject) => {
            reactRoot === null || reactRoot === void 0 ? void 0 : reactRoot.render(react_1.default.createElement(ErrorBoundary, { onRender: resolve, reportError: reject }, element));
        });
        return () => {
            reactRoot === null || reactRoot === void 0 ? void 0 : reactRoot.unmount();
            reactRootByContainer.delete(container);
        };
    }
    else {
        // react <18
        await new Promise((resolve, reject) => {
            react_dom_1.default.render(react_1.default.createElement(ErrorBoundary, { reportError: reject }, element), container, resolve);
        });
        return () => react_dom_1.default.unmountComponentAtNode(container);
    }
};
exports.reactErrorHandledRendering = reactErrorHandledRendering;
class ErrorBoundary extends react_1.default.Component {
    constructor() {
        super(...arguments);
        this.state = { hasError: false };
    }
    static getDerivedStateFromError() {
        return { hasError: true };
    }
    componentDidCatch(error, errorInfo) {
        var _a, _b;
        (_b = (_a = this.props).reportError) === null || _b === void 0 ? void 0 : _b.call(_a, error, errorInfo);
    }
    componentDidMount() {
        var _a, _b;
        (_b = (_a = this.props).onRender) === null || _b === void 0 ? void 0 : _b.call(_a);
    }
    componentDidUpdate() {
        var _a, _b;
        (_b = (_a = this.props).onRender) === null || _b === void 0 ? void 0 : _b.call(_a);
    }
    render() {
        return this.state.hasError ? null : this.props.children;
    }
}
